live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/messaging/in/mapper/RepoEventMapper.java

```diff
@@ -54,39 +63,113 @@ public class RepoEventMapper
+                calculateCustomPropertiesDelta(event));
+    }
+
+    private Long toMilliseconds(ZonedDateTime time)
+    {
+        return time == null ? null : time.toInstant().toEpochMilli();
+    }
+
+    private String getUserId(NodeResource node, Function<NodeResource, UserInfo> userInfoGetter)
+    {
+        return ofNullable(node)
+                .map(userInfoGetter)
+                .map(UserInfo::getId)
+                .orElse(null);
+    }
+
+    private <T> PropertyDelta<T> calculatePropertyDelta(RepoEvent<DataAttributes<NodeResource>> event, Function<NodeResource, T> fieldGetter)
+    {
+        if (shouldNotUpdateField(event, fieldGetter))
+        {
+            return PropertyDelta.unchanged(fieldGetter.apply(event.getData().getResource()));
+        }
+
+        return PropertyDelta.updated(fieldGetter.apply(event.getData().getResource()));
+    }
+
+    private Set<CustomPropertyDelta<?>> calculateCustomPropertiesDelta(RepoEvent<DataAttributes<NodeResource>> event)
+    {
+        if (shouldNotUpdateField(event, NodeResource::getProperties))
+        {
+            return emptySet();
+        }
+
+        if (isEventTypeCreated(event))
+        {
+            return allCustomPropertiesUpdated(event);
+        }
+
+        return someCustomPropertiesUpdated(event);
+    }
+
+    private Set<CustomPropertyDelta<?>> someCustomPropertiesUpdated(RepoEvent<DataAttributes<NodeResource>> event)
+    {
+        return customPropertiesStream(event.getData().getResourceBefore())
+                .map(Map.Entry::getKey)
+                .map(changedPropertyName -> toCustomPropertyDelta(event.getData(), changedPropertyName))
+                .collect(Collectors.toSet());
+    }
+
+    private CustomPropertyDelta<?> toCustomPropertyDelta(DataAttributes<NodeResource> event, String changedPropertyName)
+    {
+        Serializable propertyValue = event.getResource().getProperties().get(changedPropertyName);
+
+        return propertyValue == null ? CustomPropertyDelta.deleted(changedPropertyName) : CustomPropertyDelta.updated(changedPropertyName, propertyValue);
     }
 
     private Set<CustomPropertyDelta<?>> allCustomPropertiesUpdated(RepoEvent<DataAttributes<NodeResource>> event)
     {
-        return Optional.ofNullable(event.getData().getResource())
-                .map(NodeResource::getProperties)
-                .map(Map::entrySet)
-                .stream()
-                .flatMap(Collection::stream)
+        return customPropertiesStream(event.getData().getResource())
                 .filter(property -> Objects.nonNull(property.getValue()))
                 .map(property -> CustomPropertyDelta.updated(property.getKey(), property.getValue()))
                 .collect(Collectors.toSet());
     }
+
+    private Stream<Map.Entry<String, ?>> customPropertiesStream(NodeResource node)
+    {
+        return ofNullable(node)
+                .map(NodeResource::getProperties)
+                .map(Map::entrySet)
```

Comment:
Hmm... `.flatMap(Map::entrySet);` won't be enough?

Reply:
fixed